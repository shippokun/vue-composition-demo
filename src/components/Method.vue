<template>
  <button id="normalButton" @click="clickHandler">Button</button>
  <button id="complexButton" @click="complexClickHandler">Complex Button</button>
</template>

<script lang="ts">
import { defineComponent, reactive, toRefs } from 'vue';

export const createComponentStore = () => {
  const state = reactive({
    count: 0,
    flag: false,
  });
  return toRefs(state);
};
export type ComponentState = ReturnType<typeof createComponentStore>;

export const useComplexClickHandler = ({
  componentState,
}: {
  componentState: ComponentState;
}) => {
  return () => {
    console.log(componentState.count);
  };
};

export default defineComponent({
  setup() {
    const componentState = createComponentStore();

    const clickHandler = () => {
      window.alert('clicked');
    };

    const complexClickHandler = useComplexClickHandler({ componentState });

    return {
      componentState,
      clickHandler,
      complexClickHandler,
    };
  },
});
</script>
