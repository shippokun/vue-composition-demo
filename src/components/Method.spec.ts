import { mount } from '@vue/test-utils';

import Method from './Method.vue';

describe('Method.vue', () => {
  it('render', () => {
    const wrapper = mount(Method);

    expect(wrapper.html()).toMatchSnapshot();
  });
  it('button', () => {
    const wrapper = mount(Method);

    const normalButton = wrapper.find('button#normalButton');
    const mockAlert = jest.spyOn(window, 'alert').mockImplementation();
    normalButton.trigger('click');

    expect(mockAlert).toBeCalledWith('clicked');

    const complexButton = wrapper.find('button#complexButton');
    const mockLog = jest.spyOn(console, 'log').mockImplementation();
    complexButton.trigger('click');
    expect(mockLog).toBeCalled();
  });
});
