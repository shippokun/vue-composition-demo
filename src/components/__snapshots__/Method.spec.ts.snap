// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Method.vue render 1`] = `<button id="normalButton">Button</button><button id="complexButton">Complex Button</button>`;
